File: Android/app/src/androidTest/java/com/etiennebaudoux/clipboardzanager/mocks/CloudStorageProviderMock.java
package com.etiennebaudoux.clipboardzanager.mocks;

import com.etiennebaudoux.clipboardzanager.componentmodel.cloudstorage.CloudFile;
import com.etiennebaudoux.clipboardzanager.componentmodel.cloudstorage.CloudFolder;
import com.etiennebaudoux.clipboardzanager.componentmodel.cloudstorage.CloudStorageProvider;
import com.etiennebaudoux.clipboardzanager.componentmodel.cloudstorage.CloudTokenProvider;
import com.etiennebaudoux.clipboardzanager.componentmodel.core.StringUtils;
import com.etiennebaudoux.clipboardzanager.componentmodel.core.events.Event;
import com.etiennebaudoux.clipboardzanager.componentmodel.core.events.EventArgs;
import com.etiennebaudoux.clipboardzanager.componentmodel.core.tasks.Task;
import com.etiennebaudoux.clipboardzanager.componentmodel.exceptions.NotAuthenticatedException;

import java.io.FileNotFoundException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Date;

public class CloudStorageProviderMock implements CloudStorageProvider {

    private boolean _isAuthenticated;

    @Override
    public String getCloudServiceName() {
        return "MockCloudStorageProvider";
    }

    @Override
    public boolean isAuthenticated() {
        return _isAuthenticated;
    }

    @Override
    public boolean credentialExists() {
        String token = getTokenProvider().getToken("MyToken");
        return !StringUtils.isNullOrEmpty(token);
    }

    @Override
    public CloudTokenProvider getTokenProvider() {
        return new CloudTokenProviderMock();
    }

    @Override
    public Event<EventArgs> getCallbackOnResumeAfterAuthentication() {
        return null;
    }

    @Override
    public Task<Boolean> tryAuthenticateAsync() {
        return new Task<>(() -> {
            if (isAuthenticated()) {
                signOutAsync().await();
            }

            _isAuthenticated = false;
            return isAuthenticated();
        });
    }

    @Override
    public Task<Boolean> tryAuthenticateWithUiAsync(Boolean isResuming) {
        return new Task<>(() -> {
            if (isAuthenticated()) {
                signOutAsync().await();
            }

            _isAuthenticated = isResuming; //!authentication.authenticateAsync("http://authenticationUri", "http://exceptedUi").await().isCanceled();
            return isAuthenticated();
        });
    }

    @Override
    public Task<Void> signOutAsync() {
        return new Task<>(() -> {
            _isAuthenticated = false;
            getTokenProvider().setToken("MyToken", "");
            return null;
        });
    }

    @Override
    public Task<String> getUserNameAsync() {
        thowIfNotConnected();

        return new Task<>(() -> "John DOE");
    }

    @Override
    public Task<String> getUserIdAsync() {
        thowIfNotConnected();

        return new Task<>(() -> "{1df5a5312fa35f3ae1df35e1869}");
    }

    @Override
    public Task<CloudFolder> getAppFolderAsync() {
        return new Task<>(() -> {
            thowIfNotConnected();

            CloudFolder result = new CloudFolder();
            result.setFullPath("/path/MyApp");
            result.setName("MyApp");
            result.setSize(0);
            return result;
        });
    }

    @Override
    public Task<CloudFile> uploadFileAsync(InputStream baseStream, String remotePath) {
        return new Task<>(() -> {
            thowIfNotConnected();

            if (baseStream == null) {
                throw new NullPointerException();
            }

            CloudFile result = new CloudFile();
            result.setFullPath(remotePath);
            result.setName("myFile.txt");

            Date currentDate = new Date(System.currentTimeMillis());
            result.setLastModificationUtcDate(currentDate);
            return result;
        });
    }

    @Override
    public Task<Void> downloadFileAsync(String remotePath, OutputStream targetStream) {
        return new Task<>(() -> {
            thowIfNotConnected();

            if (targetStream == null) {
                throw new NullPointerException();
            }

            if (remotePath.equals("/path/MyApp/myFile.txt")) {
                throw new FileNotFoundException();
            }

            byte[] data = new byte[]{0, 1, 2, 3, 4};
            targetStream.write(data, 0, data.length);

            return null;
        });
    }

    @Override
    public Task<Void> deleteFileAsync(String remotePath) {
        return new Task<>(() -> {
            if (remotePath.equals("/path/MyApp/myFile.txt")) {
                throw new FileNotFoundException();
            }

            return null;
        });
    }

    private void thowIfNotConnected() {
        if (!isAuthenticated()) {
            throw new NotAuthenticatedException();
        }
    }
}

File: ClipboardZanager/Tests/ClipboardZanager.Shared.Tests/Mocks/CloudStorageProviderMock.cs
using System;
using System.IO;
using System.Threading.Tasks;
using ClipboardZanager.Shared.CloudStorage;
using ClipboardZanager.Shared.Exceptions;
using System.Windows.Controls;

namespace ClipboardZanager.Shared.Tests.Mocks
{
    public class CloudStorageProviderMock : ICloudStorageProvider
    {
        public string CloudServiceName => "MockCloudStorageProvider";

        /// <inheritdoc/>
        public ControlTemplate CloudServiceIcon { get; }

        public bool IsAuthenticated { get; private set; }

        public bool CredentialExists => !string.IsNullOrEmpty(TokenProvider.GetToken("MyToken"));

        public ICloudTokenProvider TokenProvider => new CloudTokenProviderMock();

        public async Task<bool> TryAuthenticateAsync()
        {
            if (IsAuthenticated)
            {
                await SignOutAsync();
            }

            IsAuthenticated = false;
            return IsAuthenticated;
        }

        public async Task<bool> TryAuthenticateWithUiAsync(ICloudAuthentication authentication)
        {
            if (IsAuthenticated)
            {
                await SignOutAsync();
            }

            IsAuthenticated = !(await authentication.AuthenticateAsync("http://authenticationUri", "http://exceptedUi")).IsCanceled;
            return IsAuthenticated;
        }

        public Task SignOutAsync()
        {
            return Task.Run(() =>
            {
                IsAuthenticated = false;
                TokenProvider.SetToken("MyToken", string.Empty);
            });
        }

        public Task<string> GetUserNameAsync()
        {
            ThowIfNotConnected();

            return Task.Run(() => "John DOE");
        }

        public Task<string> GetUserIdAsync()
        {
            ThowIfNotConnected();

            return Task.Run(() => "{1df5a5312fa35f3ae1df35e1869}");
        }

        public Task<CloudFolder> GetAppFolderAsync()
        {
            return Task.Run(() =>
            {
                ThowIfNotConnected();

                return new CloudFolder { FullPath = "/path/MyApp", Name = "MyApp", Size = 0 };
            });
        }

        public Task<CloudFile> UploadFileAsync(Stream baseStream, string remotePath)
        {
            return Task.Run(() =>
            {
                ThowIfNotConnected();

                if (baseStream == null)
                {
                    throw new NullReferenceException();
                }

                return new CloudFile
                {
                    FullPath = remotePath,
                    Name = "myFile.txt",
                    LastModificationUtcDate = DateTime.Now
                };
            });
        }

        public Task DownloadFileAsync(string remotePath, Stream targetStream)
        {
            return Task.Run(() =>
            {
                ThowIfNotConnected();

                if (targetStream == null)
                {
                    throw new NullReferenceException();
                }

                if (remotePath != "/path/MyApp/myFile.txt")
                {
                    throw new FileNotFoundException();
                }

                var data = new byte[] { 0, 1, 2, 3, 4 };
                targetStream.Write(data, 0, data.Length);
            });
        }

        public Task DeleteFileAsync(string remotePath)
        {
            if (remotePath != "/path/MyApp/myFile.txt")
            {
                throw new FileNotFoundException();
            }

            return Task.CompletedTask;
        }

        private void ThowIfNotConnected()
        {
            if (!IsAuthenticated)
            {
                throw new NotAuthenticatedException();
            }
        }
    }
}
